### DISCUSSION TITLE
How to capture tool output e.g. gcloud

### DISCUSSION THREAD

**Discussion by Kossak-esky** [xonsh/xonsh#5730] [Q&A] [answered]
This works in bash:
```bash
# in bash:
gcloud storage du -s $(gcloud storage ls gs://bucket)
```
and returns output similat to:

```
4            gs://bucket/file1.txt
60462205728  gs://bucket/dir1
1345522      gs://bucket/dir2
```

How do I do it in xonsh? I thought using `@$(...)` would yield the same result, but this command doesn't work as expected (doesn't return size of 3 above folders, only prints them and then starts to calculate size of every bucket):

```bash
# in xonsh v0.18.4
gcloud storage du -s @$(gcloud storage ls gs://bucket)
```

outptut:

```
gs://bucket/file1.txt
gs://bucket/dir1
gs://bucket/dir2
20392676     gs://bucket2
22396686     gs://bucket3
etc...
```

It looks like the main command doesn't see arguments from subcommand at all (they are printed to the terminal instead of being arguments to main command).

I use linux, xonsh/0.18.4

**Comment by anki-code** [comment]
Try to run `showcmd @$(gcloud storage ls gs://bucket)` to review the list of arguments.

**Comment by Kossak-esky** [comment]
Nice trick to debug my commands, thank you. I ran both commands and It does look like the subcommand result is just printed to terminal, without it being provided as arguments to the main command. Any idead how to fix it? Am I doing something wrong?

```
@ showcmd @$(gcloud storage ls gs://bucket)
gs://bucket/file1.txt
gs://bucket/dir1/
gs://bucket/dir2/
usage: showcmd [-h|--help|cmd args]

Displays the command and arguments as a list of strings that xonsh would
run in subprocess mode. This is useful for determining how xonsh evaluates
your commands and arguments prior to running these commands.

optional arguments:
  -h, --help            show this help message and exit

Examples
--------
  >>> showcmd echo $USER "can't" hear "the sea"
  ['echo', 'I', "can't", 'hear', 'the sea']

@ showcmd gcloud storage du -s @$(gcloud storage ls gs://bucket)
gs://bucket/file1.txt
gs://bucket/dir1/
gs://bucket/dir2/
['gcloud', 'storage', 'du', '-s']

```

**Reply by anki-code** [reply, marked as answer]
First of all please start learning from [xonsh-cheatsheet](https://github.com/anki-code/xonsh-cheatsheet) to understand xonsh better.

I think it's related to threading prediction for `gcloud`. This tool contains the part that told xonsh that it's interactive tool and threading/capturing is predicted as False.

I can repeat what you're seeing using thread/unthread decorator aliases and more simpler tools:
```xsh
xpip install xontrib-dalias  # https://github.com/anki-code/xontrib-dalias
xontrib load dalias

showcmd echo @$(echo 123)  # threaded/captured by default
# ['echo', '123']

showcmd echo @$(@unthread echo 123)  # your case
# 123
# ['echo']
```

So as described in the xonsh cheatsheet you have three ways:
1. Use decorator alias explicitly i.e. `showcmd echo @$(@thread gcloud)` - switch tool to thread mode.
2. Or set the prediction for `gcloud` i.e. `__xonsh__.commands_cache.threadable_predictors['gcloud'] = lambda *a, **kw: True`
3. Pipe the result to threadable/capturable tool i.e. `showcmd echo @$(gcloud | head)`

**Reply by Kossak-esky** [reply]
Thank you very much, also for the cheatsheet!
